internal/services/notification/notification_test.go

```diff
@@ -0,0 +1,170 @@
+// limitations under the License.
+
+package notification
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/rs/zerolog"
+	"github.com/sorintlab/errors"
+
+	"agola.io/agola/internal/services/config"
+	"agola.io/agola/internal/sqlg/sql"
+	"agola.io/agola/internal/testutil"
+	"agola.io/agola/services/notification/types"
+)
+
+func setupNotificationService(ctx context.Context, t *testing.T, log zerolog.Logger, dir string) *NotificationService {
+	dbType := testutil.DBType(t)
+	_, _, dbConnString := testutil.CreateDB(t, log, ctx, dir)
+
+	c := config.Config{
+		Notification: config.Notification{
+			DB: config.DB{
+				Type:       dbType,
+				ConnString: dbConnString,
+			},
+		},
+	}
+
+	rs, err := NewNotificationService(ctx, log, &c)
+	if err != nil {
+		t.Fatalf("unexpected err: %v", err)
+	}
+
+	return rs
+}
+
+func TestRunWebhookDelivery(t *testing.T) {
+	t.Parallel()
+
+	dir := t.TempDir()
+	ctx := context.Background()
+	log := testutil.NewLogger(t)
+
+	ns := setupNotificationService(ctx, t, log, dir)
+
+	t.Logf("starting ns")
+
+	time.Sleep(1 * time.Second)
+
+	// test delivery fail
+
+	runWebhook := createRunWebhook(t, ctx, ns)
+	runWebhookDelivery := createRunWebhookDelivery(t, ctx, ns, runWebhook.ID, types.DeliveryStatusNotDelivered)
+
+	if err := ns.handleRunWebhookDelivery(ctx, runWebhookDelivery.ID); err != nil {
+		t.Fatalf("unexpected err: %v", err)
+	}
+
+	runWebhookDeliveries := getRunWebhookDeliveries(t, ctx, ns)
+	if len(runWebhookDeliveries) != 1 {
+		t.Fatalf("expected %d runWebhook deliveries got: %d", 1, len(runWebhookDeliveries))
+	}
+	if runWebhookDeliveries[0].DeliveryStatus != types.DeliveryStatusDeliveryError {
+		t.Fatalf("expected runWebhook delivery status %q, got %q", types.DeliveryStatusDeliveryError, runWebhookDeliveries[0].DeliveryStatus)
+	}
+
+	// test delivery success
+
+	wr := setupWebhooksReceiver(ctx, t)
+	defer wr.stop()
+
+	t.Logf("starting webhooks client")
+
+	ns.c.WebhookURL = fmt.Sprintf("%s/%s", wr.exposedURL, "webhooks")
+	ns.c.WebhookSecret = "secretkey"
+
+	runWebhook = createRunWebhook(t, ctx, ns)
+	runWebhookDelivery = createRunWebhookDelivery(t, ctx, ns, runWebhook.ID, types.DeliveryStatusNotDelivered)
+
+	if err := ns.handleRunWebhookDelivery(ctx, runWebhookDelivery.ID); err != nil {
+		t.Fatalf("unexpected err: %v", err)
+	}
+
+	runWebhookDeliveries = getRunWebhookDeliveries(t, ctx, ns)
+	if len(runWebhookDeliveries) != 2 {
+		t.Fatalf("expected %d runWebhook deliveries got: %d", 2, len(runWebhookDeliveries))
+	}
+	if runWebhookDeliveries[0].DeliveryStatus != types.DeliveryStatusDeliveryError {
+		t.Fatalf("expected runWebhook delivery status %q, got %q", types.DeliveryStatusDeliveryError, runWebhookDeliveries[0].DeliveryStatus)
+	}
+	if runWebhookDeliveries[1].DeliveryStatus != types.DeliveryStatusDelivered {
+		t.Fatalf("expected runWebhook delivery status %q, got %q", types.DeliveryStatusDelivered, runWebhookDeliveries[1].DeliveryStatus)
+	}
```

Comment:
You should also check that running handleRunWebhookDelivery doesn't try the delivery is sothing already handled previously.